### PR TITLE
resolve: Avoid emitting redundant path resolution errors

### LINKED ISSUE #32760: Duplicate path resolution errors in `resolve`
For example,
```rust
mod foo {}

use foo::bar::baz::{}; // This emits 2 errors
fn main() {
    foo::bar::baz(); // This emits 2 errors
    <() as foo::bar>::baz(); // This emits 3 errors
    foo::bar::baz { x: 0 } // This emits 2 errors
}
```

### PR DESCRIPTION
**Pull request by jseyfried**
This PR avoids emitting redundant path resolution errors in `resolve` (fixes #32760).

r? @eddyb

### CODE REVIEW

File: src/test/compile-fail/issue-3521.rs [hunk 1]
```diff
@@ -15,8 +15,7 @@ fn main() {
     enum Stuff {
         Bar = foo
         //~^ ERROR attempt to use a non-constant value in a constant
-        //~| ERROR unresolved name `foo`
-        //~^^^ ERROR constant evaluation error: non-constant path in constant expression
+        //~^^ ERROR constant evaluation error: non-constant path in constant expression
```
Comment (eddyb) [inline src/test/compile-fail/issue-3521.rs]: Is the const evaluation error triggering for `DefError`? Could we suppress that, too?

File: src/test/compile-fail/macro-inner-attributes.rs [hunk 2]
```diff
@@ -26,6 +26,5 @@ test!(b,
 fn main() {
     a::bar();
     //~^ ERROR failed to resolve. Use of undeclared type or module `a`
-    //~^^ ERROR unresolved name `a::bar`
```
Comment (eddyb) [inline src/test/compile-fail/macro-inner-attributes.rs]: Nice! :heart_eyes_cat:

File: src/librustc_resolve/lib.rs [hunk 3]
```diff
@@ -1892,15 +1893,17 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
...
+                        if let PathResolution { base_def: Def::TyParam(..), .. } = path_res {
+                            Ok(self.record_def(eq_pred.id, path_res))
+                        } else {
+                            Err(false)
+                        }
+                    }).map_err(|error_reported| {
                         self.record_def(eq_pred.id, err_path_resolution());
-                    }
+                        if error_reported { return }
+                        resolve_error(self, eq_pred.span, ResolutionError::UndeclaredAssociatedType);
```
Comment (eddyb) [inline src/librustc_resolve/lib.rs]: Tidy error: `line longer than 100 chars`.

File: src/librustc_resolve/lib.rs [hunk 4]
```diff
@@ -3040,25 +3040,25 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
...
-                                err.span_help(expr.span, &msg);
-                            }
-                            err.emit();
-                        }
-                        _ => {
-                            // Keep reporting some errors even if they're ignored above.
-                            self.resolve_path(expr.id, path, 0, ValueNS);
 
+                    if let Ok(Def::Struct(..)) = type_res.map(|r| r.base_def) {
+                        let error_variant = ResolutionError::StructVariantUsedAsFunction(&path_name);
```
Comment (eddyb) [inline src/librustc_resolve/lib.rs]: Tidy error: `line longer than 100 chars`.

**Comment by eddyb** [comment]
r=me with tidy errors fixed.

File: src/test/compile-fail/issue-3521.rs (hunk 1 above)
Comment (jseyfried) [inline src/test/compile-fail/issue-3521.rs]: I believe so -- I'll look into it.

**Comment by jseyfried** [comment]
@eddyb Thanks for the quick review! I'll r=you once Travis passes.

**Pull request merged**